File: apps/api/src/scripts/run-job.ts
import "reflect-metadata";
import { Logger } from "@nestjs/common";
import { NestFactory } from "@nestjs/core";
import { AppModule } from "../app.module";
import {
  DAILY_ALERT_JOB,
  DailyAlertJob,
} from "../components/alerts/jobs/daily-alert.job";
import { JobRunner } from "../components/jobs/job-runner.service";

/**
 * Job CLI — run a scheduled job now, without waiting for its cron.
 *
 *   node dist/scripts/run-job.js daily-alert
 *
 * Goes through JobRunner exactly as the scheduler does, so a manual run is
 * recorded and instrumented identically. That matters for Milestone 2: proving
 * idempotency means re-running the real job, not a special test path.
 */
async function main() {
  const [, , jobName] = process.argv;

  if (!jobName) {
    process.stderr.write(
      `usage: run-job <job>\navailable: ${DAILY_ALERT_JOB}\n`,
    );
    process.exit(2);
  }

  const app = await NestFactory.createApplicationContext(AppModule);

  try {
    switch (jobName) {
      case DAILY_ALERT_JOB: {
        const job = app.get(DailyAlertJob);
        await app.get(JobRunner).run(DAILY_ALERT_JOB, () => job.execute());
        break;
      }
      default:
        process.stderr.write(`unknown job: ${jobName}\n`);
        process.exitCode = 2;
    }
  } finally {
    await app.close();
    Logger.flush();
  }
}

void main();

File: apps/api/src/pollen/scripts/run_job.py
"""Job CLI — run a scheduled job now, without waiting for its cron.

    pollen-job daily-alert

Goes through JobRunner exactly as the scheduler does, so a manual run is
recorded and instrumented identically. That matters for Milestone 2: proving
idempotency means re-running the real job, not a special test path.
"""

from __future__ import annotations

import asyncio
import logging
import sys

from pollen.components.alerts.jobs.daily_alert import DAILY_ALERT_JOB
from pollen.container import build_container
from pollen.db import connect

logging.basicConfig(level=logging.INFO, format="%(levelname)s %(name)s: %(message)s")

JOBS = {DAILY_ALERT_JOB}


async def _run(job_name: str) -> int:
    container = build_container()
    client = await connect(container.settings.mongodb_uri)
    try:
        await container.runner.run(job_name, container.daily_alert.execute)
    finally:
        await client.close()
    return 0


def main() -> int:
    if len(sys.argv) < 2:
        sys.stderr.write(f"usage: pollen-job <job>\navailable: {', '.join(sorted(JOBS))}\n")
        return 2

    job_name = sys.argv[1]
    if job_name not in JOBS:
        sys.stderr.write(f"unknown job: {job_name}\n")
        return 2

    return asyncio.run(_run(job_name))


if __name__ == "__main__":
    raise SystemExit(main())
